### PR TITLE
feat: notifications library  - [2 of 10]

### PR DESCRIPTION
**Pull request by Jonathansoufer**
## **Description**

Within the scope of [Notifications Project](https://github.com/MetaMask/mobile-planning/issues/1495), this PR replaces libraries and approaches on how MetaMask handles notifications to our users. The adoption of this approach will pave the way for MM make usage of push notifications (back/foregrounds).

REMINDER: Although this PR can be reviewed separately, triggering a NEW BUILD should be done together with [UI changes PR](https://github.com/MetaMask/metamask-mobile/pull/8961) so the user can have full control about push notifications.

## **Related issues**

Fixes: [Jira Ticket](https://consensyssoftware.atlassian.net/browse/NOTIFY-466?atlOrigin=eyJpIjoiOWM0YTQ1YjVlN2UzNGFkZWIyZWY0M2ExZDE4ODE3NTMiLCJwIjoiaiJ9)

## **Manual testing steps**

1. Go to any token details page.
2. Submit a transaction and close the app.
3. Wait for the Push Notification and click on it.

### **After**

https://github.com/MetaMask/metamask-mobile/assets/44679989/c64f88a3-1613-42a8-a076-1aecc266413a

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/components/Views/Root/index.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-console */
```
Comment (owencraston) [inline app/components/Views/Root/index.js:1]: is this still needed?

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -321,9 +321,9 @@
 				650F2B9D24DC5FF200C3B9C4 /* libRCTAesForked.a in Frameworks */,
 				153C1ABB2217BCDC0088EFE0 /* JavaScriptCore.framework in Frameworks */,
 				153F84CA2319B8FD00C19B63 /* Branch.framework in Frameworks */,
-				0FD509E0336BF221F6527B24 /* BuildFile in Frameworks */,
-				D45BF85DECACCB74EDCBE88A /* BuildFile in Frameworks */,
-				B638844E306CAE9147B52C85 /* BuildFile in Frameworks */,
+				0FD509E0336BF221F6527B24 /* (null) in Frameworks */,
```
Comment (owencraston) [inline ios/MetaMask.xcodeproj/project.pbxproj:324]: why are these all null?

File: package.json [hunk 3]
```diff
@@ -289,7 +290,6 @@
     "react-native-os": "^1.2.6",
     "react-native-permissions": "^3.7.2",
     "react-native-progress": "3.5.0",
-    "react-native-push-notification": "git+https://github.com/MetaMask/react-native-push-notification.git#975e0e6757c40e3dce2d1b6dd3d66fc91127ac4c",
```
Comment (owencraston) [inline package.json:292]: 👏

File: app/util/notifications/types.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+export const NotificationTypes = {
+  transaction: 'transaction',
+  marketing: 'marketing',
+} as const;
+
+export type NotificationTypesType =
+  (typeof NotificationTypes)[keyof typeof NotificationTypes];
+
+export const NotificationTransactionTypes = {
```
Comment (owencraston) [inline app/util/notifications/types.ts:9]: do we control these type or do they map to an internal type from the notifee library?

File: app/util/notifications/settings/storage/index.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { MMKV } from 'react-native-mmkv';
+import { STORAGE_TYPES, STORAGE_IDS, mapStorageTypeToIds } from './constants';
+
+export const notificationSettingsStorage = new MMKV({
+  id: STORAGE_IDS.NOTIFICATIONS,
+});
+
+export default class mmStorage {
```
Comment (owencraston) [inline app/util/notifications/settings/storage/index.ts:8]: this is great. we should definitly migrate all of our storage to use this class cc @Cal-L

File: app/core/NotificationManager.js [hunk 6]
```diff
@@ -433,7 +460,7 @@ class NotificationManager {
             assetType: strings('unit.eth'),
           },
           autoHide: true,
-          duration: 5000,
+          duration: 7000,
```
Comment (owencraston) [inline app/core/NotificationManager.js:463]: any specific reason for this increase?

File: app/components/Nav/Main/index.js [hunk 7]
```diff
@@ -339,6 +332,42 @@ const Main = (props) => {
     termsOfUse();
   }, [termsOfUse]);
 
+  useEffect(() => {
```
Comment (owencraston) [inline app/components/Nav/Main/index.js:335]: lets extract this useEffect into its own dedicated hook similarly to how we did with...
```jsx
useEnableAutomaticSecurityChecks();
  useMinimumVersions();
```

Comment (sethkfman) [reply app/components/Nav/Main/index.js:335]: Yes and add tests to this extracted hook.

File: app/util/notifications/settings/storage/constants.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+export const STORAGE_TYPES = {
+  STRING: 'string',
+  BOOLEAN: 'boolean',
+  NUMBER: 'number',
+  OBJECT: 'object',
+};
+
+// Map all non string storage ids to their respective types
+export const mapStorageTypeToIds = (id: string) => {
```
Comment (sethkfman) [inline app/util/notifications/settings/storage/constants.ts:24]: Add unit tests

File: app/core/NotificationManager.js [hunk 9]
```diff
@@ -20,54 +19,54 @@ import { store } from '../store';
 const constructTitleAndMessage = (data) => {
```
Comment (sethkfman) [inline app/core/NotificationManager.js:19]: Add tests for this manager file

File: ios/MetaMask/Info.plist [hunk 10]
```diff
@@ -134,6 +144,8 @@
 		<key>test</key>
 		<string>$(MM_BRANCH_KEY_TEST)</string>
 	</dict>
+	<key>mixpanel_token</key>
+	<string>$(MM_MIXPANEL_TOKEN)</string>
```
Comment (sethkfman) [inline ios/MetaMask/Info.plist:148]: This token should be removed. @NicolasMassart Can you confirm?

File: app/components/Views/Root/index.js (hunk 1 above)
Comment (Jonathansoufer) [reply app/components/Views/Root/index.js:1]: not anymore. it was a leftover. Good catch. Fixed.

File: ios/MetaMask.xcodeproj/project.pbxproj (hunk 2 above)
Comment (Jonathansoufer) [reply ios/MetaMask.xcodeproj/project.pbxproj]: that's fixed already.

File: app/util/notifications/types.ts (hunk 4 above)
Comment (Jonathansoufer) [reply app/util/notifications/types.ts:9]: All these types are controlled by us with no mapping to notifee whatsoever. These are types we aim to use both for transactions interactions and marketing (aka. FCM)

File: app/core/NotificationManager.js [hunk 11]
```diff
@@ -275,20 +292,25 @@ class NotificationManager {
...
-      PUSH_NOTIFICATIONS_PROMPT_COUNT,
-    );
-    if (
-      !promptCount ||
-      Number(promptCount) < AppConstants.MAX_PUSH_NOTIFICATION_PROMPT_TIMES
-    ) {
-      PushNotification.checkPermissions((permissions) => {
-        if (!permissions || !permissions.alert) {
+    try {
+      const promptCount = mmStorage.getLocal(
```
[the remark this answers is not in the document]
Comment (Jonathansoufer) [reply app/core/NotificationManager.js:301]: Firebase will be used later integrated to allow a) Consensys broadcast notifications (remote) to MM wallets only when MM auth is integrated and b) to profile sync, again heavily dependent on MM auth. Either way, we need to store locally user's choice `STORAGE_IDS.PUSH_NOTIFICATIONS_PROMPT_COUNT` to not bother our users if they said they don't wanna get notified.

File: app/core/NotificationManager.js (hunk 6 above)
Comment (Jonathansoufer) [reply app/core/NotificationManager.js:463]: based on our assumptions/observations, 5000 was too quick to allow users read notifications properly. Not a big deal if we wanna rollback this value.

File: app/util/notifications/settings/storage/constants.ts (hunk 8 above)
Comment (Jonathansoufer) [reply app/util/notifications/settings/storage/constants.ts:24]: added

File: app/components/Nav/Main/index.js (hunk 7 above)
Comment (Jonathansoufer) [reply app/components/Nav/Main/index.js]: added base tests.

File: app/util/test/testSetup.js [hunk 12]
```diff
@@ -172,6 +172,18 @@ NativeModules.RNCNetInfo = {
   getCurrentState: jest.fn(() => Promise.resolve()),
 };
 
+NativeModules.RCTAnalytics = {
+  optIn: jest.fn(),
+  trackEvent: jest.fn(),
+  getRemoteVariables: jest.fn(),
+};
```
Comment (NicolasMassart) [inline app/util/test/testSetup.js:179]: We removed this from the app, now the analytics are handled by Segment SDK.

Comment (Jonathansoufer) [reply app/util/test/testSetup.js:179]: Removed.

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me.

File: package.json [hunk 13]
```diff
@@ -170,8 +170,10 @@
     "@metamask/transaction-controller": "8.0.1",
     "@metamask/utils": "^8.1.0",
     "@ngraveio/bc-ur": "^1.1.6",
+    "@notifee/react-native": "^7.8.2",
     "@react-native-async-storage/async-storage": "1.17.10",
     "@react-native-clipboard/clipboard": "1.8.4",
+    "@react-native-community/async-storage": "^1.12.1",
```
Comment (owencraston) [inline package.json:176]: why are we using this when we could be using MMKV which is 1000 times faster than async storage?

Comment (owencraston) [reply package.json:176]: it doesn't seem to even be used in this PR.

Comment (Jonathansoufer) [reply package.json:176]: This is a dependency of `asyncstorage-down`, which before patch [asyncstorage-down+4.2.0.patch](https://github.com/MetaMask/metamask-mobile/blob/main/patches/asyncstorage-down+4.2.0.patch) was using deprecated `react-native').AsyncStorage` and after the patch, uses `@react-native-community/async-storage').default`.

**Comment by Jonathansoufer** [comment]
> @Jonathansoufer You will probably have to deal with Apple Privacy Manifest when integrating Firebase. Review [this](https://github.com/MetaMask/metamask-mobile/pull/9124) PR for how it was addressed.

@sethkfman thanks for the heads up. Based on `https://firebase.google.com/docs/ios/app-store-data-collection` we'll absolutely have to update our manifest, but since this PR doesn't implement FCM yet, I'll handle it in a separated PR.

**Review by owencraston — APPROVED** [review]

**Pull request merged**